uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/implementation/addon/MqttMessageEvent.java

```diff
@@ -0,0 +1,114 @@
+     *
+     * @param connectionControl the connection control which receives that message
+     * @param message the gRPC presentation of received MQTT message
+     */
+    public MqttMessageEvent(@NonNull ConnectionControl connectionControl, @NonNull Mqtt5Message message) {
+        super(Type.EVENT_TYPE_MQTT_MESSAGE);
+        this.connectionControl = connectionControl;
+        this.message = message;
+    }
+
+    @Override
+    public String getConnectionName() {
+        return connectionControl.getConnectionName();
+    }
+
+    @Override
+    public boolean isMatched(@NonNull EventFilter filter) {
+        // check type and timestamp
+        boolean matched = super.isMatched(filter);
+        if (!matched) {
+            return false;
+        }
+
+        // check connection
+        matched = compareConnection(filter.getConnectionControl(), filter.getAgentId(), filter.getConnectionId(), 
+                                    filter.getConnectionName());
+        if (!matched) {
+            return false;
+        }
+
+        // check topic
+        final String topic = filter.getTopic();
+        if (topic != null) {
+            if (!topic.equals(message.getTopic())) {
+                return false;
+            }
+        } else {
+            // or topic filter
+            final String topicFilter = filter.getTopicFilter();
+            if (topicFilter != null && !isTopicMatched(message.getTopic(), topicFilter)) {
+                return false;
+            }
+        }
+
+        // TODO: check QoS ?
+
+        // check content
+        return comparePayload(filter.getContent());
+    }
+
+    @SuppressWarnings("PMD.CompareObjectsWithEquals")
+    private boolean compareConnection(ConnectionControl expectedConnectionControl, String agentId, Integer connectionId,
+                                        String connectionName) {
+        // 1'st priority
+        if (expectedConnectionControl != null) {
+            // compare references !
+            return expectedConnectionControl == connectionControl;
+        }
+
+        // 2'nd priority
+        if (agentId != null && connectionId != null) {
+            return agentId.equals(connectionControl.getAgentControl().getAgentId())
+                        && connectionId == connectionControl.getConnectionId();
+        }
+
+        // 3'th priority
+        // check connection name
+        return connectionName == null || connectionName.equals(getConnectionName());
+    }
+
+    private boolean comparePayload(byte[] expected) {
+        if (expected == null) {
+            return true;
+        }
+
+        ByteString byteStringPayload = message.getPayload();
+        if (byteStringPayload == null) {
+            return false;
+        } else {
+            return Arrays.equals(expected, byteStringPayload.toByteArray());
+        }
+    }
+
+    private static boolean isTopicMatched(@NonNull String topic, @NonNull String topicFilter) {
```

Comment:
You can steal the implementation from the Nucleus:
https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/94a91fb3344a9069462d7bb8828d1376304ec67b/src/main/java/com/aws/greengrass/mqttclient/MqttTopic.java#L44

Reply:
Yes please copy the implementation into a utility class for the mqtt topics, then use the implementation here.

Reply:
Added in fed30933e0c81aa468fdcd2bd25fb21bdd08cfc0